src/Engines/Nunjucks.js

```diff
@@ -4,6 +4,78 @@ const TemplatePath = require("../TemplatePath");
 const EleventyErrorUtil = require("../EleventyErrorUtil");
 const EleventyBaseError = require("../EleventyBaseError");
 
+/*
+ * HACKITYHACKHACKHACKHACK
+ */
+let inc = ((i) => {
+  return () => {
+    return i++;
+  };
+})(0);
+let pathMap = new Map();
+
+let clearCache = (reason = "") => {
+  if (pathMap.size > 0) {
+    pathMap.clear();
+  }
+};
+
+let clearAndWrap = (name, obj = NunjucksLib.Environment) => {
+  let orig = obj.prototype[name];
+  obj.prototype[name] = function (...args) {
+    clearCache(name);
+    return orig.call(this, ...args);
+  };
+};
+
+clearAndWrap("addExtension");
+clearAndWrap("removeExtension");
+
+(function () {
+  // IFFE to prevent temp var leakage
+  let getKey = (obj) => {
+    let eids = obj.env.extensionsList.reduce((v, c) => {
+      return c.__id ? v + " : " + c.__id : v;
+    }, "");
+    let k =
+      `${obj.path || obj.tmplStr} :: ${obj.tmplStr.length} :: ` +
+      `${obj.env.asyncFilters.length} :: (${eids})`; /* +
+            (
+              (obj.env.extensionsList.length) ? 
+                `${Object.keys(obj.env.extensions).join("|")} :: (${eids})` :
+                ""
+            );
+            */
+    return k;
+  };
+  let tc = NunjucksLib.Template.prototype._compile;
+  NunjucksLib.Template.prototype._compile = function (...args) {
+    // console.log(`NunjucksLib.Template.prototype._compile`);
+    if (!this.compiled && !this.tmplProps && pathMap.has(getKey(this))) {
+      let pathProps = pathMap.get(getKey(this));
+      this.blocks = pathProps.blocks;
+      this.rootRenderFunc = pathProps.rootRenderFunc;
+      this.compiled = true;
+    } else {
+      tc.call(this, ...args);
+      pathMap.set(getKey(this), {
+        blocks: this.blocks,
+        rootRenderFunc: this.rootRenderFunc,
+      });
+    }
+  };
+
+  let ae = NunjucksLib.Environment.prototype.addExtension;
+  NunjucksLib.Environment.prototype.addExtension = function (...args) {
+    let e = args[1];
```

Comment:
Normally, what is the second argument passed into the extension?
The body?
What if you spread out by letting the first argument be assigned to `name` and the second to `implementation`?

Reply:
yeah, sorry, that was lazy; will rework.

Reply:
Fixed in updated patch.